### PR TITLE
Fix TV commissioning and control

### PR DESCRIPTION
**Pull request by chrisdecenzo**
Two items fixed in this PR
1. Recent changes to default operational discovery advertising for commissioners broke casting (when restarting client)
2. Recent UDC changes did not allow optional pid in new vidList

### CODE REVIEW

**Review by cliffamzn — APPROVED** [review]

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp [hunk 1]
```diff
@@ -242,25 +242,56 @@ CHIP_ERROR IdentificationDeclaration::ReadPayload(uint8_t * udcPayload, size_t p
...
-                        err = reader.Get(mAppVendorIds[mNumAppVendorIds]);
-                        mNumAppVendorIds++;
+                        ReturnErrorOnFailure(reader.EnterContainer(outerContainerType));
+                        uint16_t appVendorId  = 0;
+                        uint16_t appProductId = 0;
+
+                        while ((err = reader.Next()) == CHIP_NO_ERROR)
+                        {
+                            containerTag = reader.GetTag();
+                            tagNum       = static_cast<uint8_t>(chip::TLV::TagNumFromTag(containerTag));
```
Comment (bzbarsky-apple) [inline src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp:265]: You can't do TagNumFromTag without checking the tag type first.  You might get garbage data....

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp [hunk 2]
```diff
@@ -242,25 +242,56 @@ CHIP_ERROR IdentificationDeclaration::ReadPayload(uint8_t * udcPayload, size_t p
...
             // app vendor list
             {
+                ChipLogError(AppServer, "TLV found an applist");
                 chip::TLV::TLVType listContainerType = chip::TLV::kTLVType_List;
                 ReturnErrorOnFailure(reader.EnterContainer(listContainerType));
 
                 while ((err = reader.Next()) == CHIP_NO_ERROR && mNumAppVendorIds < sizeof(mAppVendorIds))
                 {
                     containerTag = reader.GetTag();
                     tagNum       = static_cast<uint8_t>(chip::TLV::TagNumFromTag(containerTag));
```
Comment (bzbarsky-apple) [inline src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp:255]: Pre-existing, but TagNumFromTag can return garbage if the tag type has not been checked.

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp [hunk 3]
```diff
@@ -242,25 +242,56 @@ CHIP_ERROR IdentificationDeclaration::ReadPayload(uint8_t * udcPayload, size_t p
...
+                        }
+                        if (err == CHIP_END_OF_TLV)
+                        {
+                            ChipLogProgress(AppServer, "TLV end of struct TLV");
+                            ReturnErrorOnFailure(reader.ExitContainer(outerContainerType));
+                        }
+                        if (appVendorId != 0)
+                        {
+                            mAppVendorIds[mNumAppVendorIds] = appVendorId;
+                            mNumAppVendorIds++;
```
Comment (bzbarsky-apple) [inline src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp:284]: Isn't this an out-of-bounds write if the TLV has more than kMaxAppVendorIds things in it?

File: src/protocols/user_directed_commissioning/UserDirectedCommissioning.h [hunk 4]
```diff
@@ -256,13 +256,16 @@ class DLL_EXPORT IdentificationDeclaration
     {
         kVendorIdTag = 1,
         kProductIdTag,
-        kNameTag,
-        kRotatingIdTag,
-        kCdPortTag,
+        kDeviceNameTag,
+        kDeviceTypeTag,
         kPairingInstTag,
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/protocols/user_directed_commissioning/UserDirectedCommissioning.h:261]: this feature was just introduced and we are not yet at the point where we can't introduce breaking changes

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp (hunk 1 above)
Comment (chrisdecenzo) [reply src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp]: added checks

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp]: added checks

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp [hunk 5]
```diff
@@ -242,25 +242,56 @@ CHIP_ERROR IdentificationDeclaration::ReadPayload(uint8_t * udcPayload, size_t p
...
+                            containerTag = reader.GetTag();
+                            tagNum       = static_cast<uint8_t>(chip::TLV::TagNumFromTag(containerTag));
+                            if (tagNum == kAppVendorIdTag)
+                            {
+                                err = reader.Get(appVendorId);
+                            }
+                            else if (tagNum == kAppProductIdTag)
+                            {
+                                err = reader.Get(appProductId);
+                                ChipLogError(AppServer, "unsupported product id tag %d", tagNum);
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp]: fixed

File: src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp (hunk 3 above)
Comment (chrisdecenzo) [reply src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp]: the outer loop performs the boundary check

**Review by nicelyjust — APPROVED** [review]

**Review by younghak-hwang — APPROVED** [review]

**Pull request merged**